// fuse/dist/adapters/node.mjs
// src/adapters/node.ts
import http from "http";
import { createYoga } from "graphql-yoga";
import { builder } from "fuse";

// src/utils/yoga-helpers.ts
import { blockFieldSuggestionsPlugin } from "@escape.tech/graphql-armor-block-field-suggestions";
import { useDeferStream } from "@graphql-yoga/plugin-defer-stream";
import { useDisableIntrospection } from "@graphql-yoga/plugin-disable-introspection";
import { createStellateLoggerPlugin } from "stellate/graphql-yoga";
var getYogaPlugins = (stellate) => {
  return [
    useDeferStream(),
    process.env.NODE_ENV === "production" && useDisableIntrospection(),
    process.env.NODE_ENV === "production" && blockFieldSuggestionsPlugin(),
    Boolean(process.env.NODE_ENV === "production" && stellate) && createStellateLoggerPlugin({
      serviceName: stellate.serviceName,
      token: stellate.loggingToken,
      fetch
    })
  ].filter(Boolean);
};
var wrappedContext = (context) => {
  return async (ct) => {
    const baseContext = {
      request: ct.request,
      headers: ct.request.headers,
      params: ct.params
    };
    if (typeof context === "function") {
      const userCtx = context(baseContext);
      if (userCtx.then) {
        const result = await userCtx;
        return {
          ...baseContext,
          ...result
        };
      }
      return {
        ...baseContext,
        ...userCtx
      };
    } else if (typeof context === "object") {
      return {
        ...baseContext,
        ...context
      };
    }
    return baseContext;
  };
};

// src/adapters/node.ts
async function main() {
  let ctx;
  import.meta.glob("/types/**/*.ts", { eager: true });
  const context = import.meta.glob("/_context.ts", { eager: true });
  if (context["/_context.ts"]) {
    const mod = context["/_context.ts"];
    if (mod.getContext) {
      ctx = mod.getContext;
    }
  }
  const completedSchema = builder.toSchema({});
  const yoga = createYoga({
    graphiql: false,
    maskedErrors: true,
    schema: completedSchema,
    // We allow batching by default
    batching: true,
    context: wrappedContext(ctx),
    plugins: getYogaPlugins()
  });
  const server = http.createServer(yoga);
  server.listen(process.env.PORT || 4e3);
}
main();
export {
  main
};
